base/fitness_model.py

```diff
@@ -234,14 +242,35 @@ def calc_node_frequencies(self):
             for time in self.timepoints[:-1]:
                 node.delta_freqs[time] = np.asscalar(interpolation(time + self.delta_time))
 
+        # Estimate frequencies for tips at specific timepoints.
+        # Censor future tips from estimations unless these data are explicitly allowed.
         # freq_arrays list *all* tips for each initial timepoint
         self.freq_arrays={}
+        frequency_parameters = self.frequencies.get_params()
         for time in self.timepoints:
             tmp_freqs = []
-            for tip in self.tips:
-                tmp_freqs.append(tip.timepoint_freqs[time])
-            self.freq_arrays[time] = np.array(tmp_freqs)
 
+            if self.censor_frequencies:
+                # Censor frequencies by omitting all tips sampled after the current timepoint.
+                sys.stderr.write("Calculating censored frequencies for %s\n" % time)
+                frequency_parameters["max_date"] = time
+                frequency_estimator = type(self.frequencies)(**frequency_parameters)
```

Comment:
this currently only works with the `KDEfrequencyClass`. This is pretty opaque. Why not all KDE frequencies explicitly. In the future, we can abstract the frequencies from the underlying model such that all models work.